### PR TITLE
Fix empty RangeFilter

### LINKED ISSUE #1899: Join-Order Query 16a segfaults
```
[master]> ./build-release/hyriseBenchmarkJoinOrder -t 1
[...]
- Benchmarking 16a
Segmentation fault
```

```
(gdb) bt
#0  0x0000000002f25b7c in opossum::RangeFilter<int>::does_not_contain(opossum::PredicateCondition, boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<6l>, opossum::NullValue, boost::mpl::l_item<mpl_::long_<5l>, int, boost::mpl::l_item<mpl_::long_<4l>, long, boost::mpl::l_item<mpl_::long_<3l>, float, boost::mpl::l_item<mpl_::long_<2l>, double, boost::mpl::l_item<mpl_::long_<1l>, std::__cxx11::basic_string<char, std::char_traits<char>, boost::container::pmr::polymorphic_allocator<char> >, boost::mpl::l_end> > > > > > >> const&, std::optional<boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<6l>, opossum::NullValue, boost::mpl::l_item<mpl_::long_<5l>, int, boost::mpl::l_item<mpl_::long_<4l>, long, boost::mpl::l_item<mpl_::long_<3l>, float, boost::mpl::l_item<mpl_::long_<2l>, double, boost::mpl::l_item<mpl_::long_<1l>, std::__cxx11::basic_string<char, std::char_traits<char>, boost::container::pmr::polymorphic_allocator<char> >, boost::mpl::l_end> > > > > > >> > const&) const (this=0x7ffff703fb10,
    predicate_condition=opossum::PredicateCondition::LessThanEquals, variant_value=...,
    variant_value2=std::optional<boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<6>, opossum::NullValue, boost::mpl::l_item<mpl_::long_<5>, int, boost::mpl::l_item<mpl_::long_<4>, long, boost::mpl::l_item<mpl_::long_<3>, float, boost::mpl::l_item<mpl_::long_<2>, double, boost::mpl::l_item<mpl_::long_<1>, std::__cxx11::basic_string<char, std::char_traits<char>, boost::container::pmr::polymorphic_allocator<char> >, boost::mpl::l_end> > > > > > >>> [no contained value])
    at ../src/lib/statistics/statistics_objects/range_filter.cpp:223
#1  0x0000000002f25ea2 in opossum::RangeFilter<int>::does_not_contain(opossum::PredicateCondition, boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<6l>, opossum::NullValue, boost::mpl::l_item<mpl_::long_<5l>, int, boost::mpl::l_item<mpl_::long_<4l>, long, boost::mpl::l_item<mpl_::long_<3l>, float, boost::mpl::l_item<mpl_::long_<2l>, double, boost::mpl::l_item<mpl_::long_<1l>, std::__cxx11::basic_string<char, std::char_traits<char>, boost::container::pmr::polymorphic_allocator<char> >, boost::mpl::l_end> > > > > > >> const&, std::optional<boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<6l>, opossum::NullValue, boost::mpl::l_item<mpl_::long_<5l>, int, boost::mpl::l_item<mpl_::long_<4l>, long, boost::mpl::l_item<mpl_::long_<3l>, float, boost::mpl::l_item<mpl_::long_<2l>, double, boost::mpl::l_item<mpl_::long_<1l>, std::__cxx11::basic_string<char, std::char_traits<char>, boost::container::pmr::polymorphic_allocator<char> >, boost::mpl::l_end> > > > > > >> > const&) const (this=0x7ffff703fb10,
    predicate_condition=opossum::PredicateCondition::BetweenUpperExclusive, variant_value=...,
    variant_value2=std::optional<boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<6>, opossum::NullValue, boost::mpl::l_item<mpl_::long_<5>, int, boost::mpl::l_item<mpl_::long_<4>, long, boost::mpl::l_item<mpl_::long_<3>, float, boost::mpl::l_item<mpl_::long_<2>, double, boost::mpl::l_item<mpl_::long_<1>, std::__cxx11::basic_string<char, std::char_traits<char>, boost::container::pmr::polymorphic_allocator<char> >, boost::mpl::l_end> > > > > > >>> = {...})
    at ../src/lib/statistics/statistics_objects/range_filter.cpp:255
#2  0x0000000002e27c87 in opossum::ChunkPruningRule::_can_prune(opossum::BaseAttributeStatistics const&, opossum::PredicateCondition, boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<6l>, opossum::NullValue, boost::mpl::l_item<mpl_::long_<5l>, int, boost::mpl::l_item<mpl_::long_<4l>, long, boost::mpl::l_item<mpl_::long_<3l>, float, boost::mpl::l_item<mpl_::long_<2l>, double, boost::mpl::l_item<mpl_::long_<1l>, std::__cxx11::basic_string<char, std::char_traits<char>, boost::container::pmr::polymorphic_allocator<char> >, boost::mpl::l_end> > > > > > >> const&, std::optional<boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<6l>, opossum::NullValue, boost::mpl::l_item<mpl_::long_<5l>, int, boost::mpl::l_item<mpl_::long_<4l>, long, boost::mpl::l_item<mpl_::long_<3l>, float, boost::mpl::l_item<mpl_::long_<2l>, double, boost::mpl::l_item<mpl_::long_<1l>, std::__cxx11::basic_string<char, std::char_traits<char>, boost::container::pmr::polymorphic_allocator<char> >, boost::mpl::l_end> > > > > > >> > const&)::$_1::operator()<boost::hana::basic_type<int> >(boost::hana::basic_type<int>) const (this=0x7ffed35f80b0, data_type_t=...)
    at ../src/lib/optimizer/strategy/chunk_pruning_rule.cpp:183
```

```
(gdb) f 0
#0  0x0000000002f25b7c in opossum::RangeFilter<int>::does_not_contain(opossum::PredicateCondition, boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<6l>, opossum::NullValue, boost::mpl::l_item<mpl_::long_<5l>, int, boost::mpl::l_item<mpl_::long_<4l>, long, boost::mpl::l_item<mpl_::long_<3l>, float, boost::mpl::l_item<mpl_::long_<2l>, double, boost::mpl::l_item<mpl_::long_<1l>, std::__cxx11::basic_string<char, std::char_traits<char>, boost::container::pmr::polymorphic_allocator<char> >, boost::mpl::l_end> > > > > > >> const&, std::optional<boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<6l>, opossum::NullValue, boost::mpl::l_item<mpl_::long_<5l>, int, boost::mpl::l_item<mpl_::long_<4l>, long, boost::mpl::l_item<mpl_::long_<3l>, float, boost::mpl::l_item<mpl_::long_<2l>, double, boost::mpl::l_item<mpl_::long_<1l>, std::__cxx11::basic_string<char, std::char_traits<char>, boost::container::pmr::polymorphic_allocator<char> >, boost::mpl::l_end> > > > > > >> > const&) const (this=0x7ffff703fb10,
    predicate_condition=opossum::PredicateCondition::LessThanEquals, variant_value=...,
    variant_value2=std::optional<boost::variant<boost::detail::variant::over_sequence<boost::mpl::l_item<mpl_::long_<6>, opossum::NullValue, boost::mpl::l_item<mpl_::long_<5>, int, boost::mpl::l_item<mpl_::long_<4>, long, boost::mpl::l_item<mpl_::long_<3>, float, boost::mpl::l_item<mpl_::long_<2>, double, boost::mpl::l_item<mpl_::long_<1>, std::__cxx11::basic_string<char, std::char_traits<char>, boost::container::pmr::polymorphic_allocator<char> >, boost::mpl::l_end> > > > > > >>> [no contained value])
    at ../src/lib/statistics/statistics_objects/range_filter.cpp:223
223	      return value < min;
(gdb) p value
$3 = 100
(gdb) p min
$4 = (const int &) <error reading variable>
(gdb) p ranges
$5 = std::vector of length 0, capacity 0
```

### PR DESCRIPTION
**Pull request by mrks**
fixes #1899

### CODE REVIEW

File: scripts/test/hyriseBenchmarkJoinOrder_test.py [hunk 1]
```diff
@@ -75,10 +75,9 @@ def main():
 
   arguments = {}
   arguments["--table_path"] = "'resources/test_data/imdb_sample/'"
-  arguments["--queries"] = "'3b'"
   arguments["--time"] = "10"
-  arguments["--runs"] = "100"
-  arguments["--warmup"] = "10"
+  arguments["--runs"] = "2"
```
Comment (mrks) [inline scripts/test/hyriseBenchmarkJoinOrder_test.py]: Now executing the entire JOB, albeit on a very small data sample

File: src/lib/scheduler/abstract_task.cpp [hunk 2]
```diff
@@ -103,6 +103,11 @@ void AbstractTask::_on_predecessor_done() {
     auto worker = Worker::get_this_thread_worker();
 
     if (worker) {
+      // If the first task was executed faster than the other calls were scheduled, we might end up in a situation where
```
Comment (Bouncner) [inline src/lib/scheduler/abstract_task.cpp]: What are "other calls" here?

Comment (mrks) [reply src/lib/scheduler/abstract_task.cpp]: > If the first task was executed faster than the other **tasks** were scheduled, the successor is not properly scheduled yet. At the time of writing, this did not make a difference, but for the sake of a clearly defined life cycle, we wait for the task to be scheduled.

Oupsie. Is that better?

Comment (Bouncner) [reply src/lib/scheduler/abstract_task.cpp]: Jep.

**Comment by Bouncner** [comment]
So the summary is: we tried to create range filters for empty (as all-is-null) columns?

**Comment by mrks** [comment]
> So the summary is: we tried to create range filters for empty (as all-is-null) columns?

s/columns/segments/

Yes. While that in itself is not bad, the internal workings never expected the ranges to be empty. Instead of fixing that, I decided to make empty filters illegal.

**Pull request merged**